server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/event/PreComputeMessagePreviewListener.java

```diff
@@ -0,0 +1,102 @@
+import org.apache.james.mailbox.MessageIdManager;
+import org.apache.james.mailbox.events.Event;
+import org.apache.james.mailbox.events.Group;
+import org.apache.james.mailbox.events.MailboxListener;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.FetchGroupImpl;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.MessageResult;
+import org.parboiled.common.ImmutableList;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+
+import reactor.core.publisher.Mono;
+
+public class PreComputeMessagePreviewListener implements MailboxListener.GroupMailboxListener {
+    public static class PreComputeMessagePreviewListenerGroup extends Group {
+
+    }
+
+    private static final Group GROUP = new PreComputeMessagePreviewListenerGroup();
+
+    private final MessageIdManager messageIdManager;
+    private final MessagePreviewStore messagePreviewStore;
+    private final MessageFullViewFactory messageFullViewFactory;
+
+    private MailboxManager mailboxManager;
+
+    @Inject
+    public PreComputeMessagePreviewListener(MailboxManager mailboxManager, MessageIdManager messageIdManager,
+                                            MessagePreviewStore messagePreviewStore, MessageFullViewFactory messageFullViewFactory) {
+        this.mailboxManager = mailboxManager;
+        this.messageIdManager = messageIdManager;
+        this.messagePreviewStore = messagePreviewStore;
+        this.messageFullViewFactory = messageFullViewFactory;
+    }
+
+    @Override
+    public Group getDefaultGroup() {
+        return GROUP;
+    }
+
+    @Override
+    public void event(Event event) throws MailboxException {
+        if (event instanceof Added) {
+            MailboxSession session = mailboxManager.createSystemSession(event.getUsername());
+            handleAddedEvent((Added) event, session);
+        }
+    }
+
+    private void handleAddedEvent(Added addedEvent, MailboxSession session) throws MailboxException {
+        List<MessageId> messageIds = addedEvent.getUids()
+            .stream()
+            .map(uid -> addedEvent.getMetaData(uid).getMessageId())
+            .distinct()
+            .collect(Guavate.toImmutableList());
+
+        preComputePreview(messageIds, session);
+    }
+
+    private void preComputePreview(List<MessageId> messagesId, MailboxSession session) throws MailboxException {
+        messageIdManager.getMessages(messagesId, FetchGroupImpl.BODY_CONTENT, session)
+            .stream()
+            .map(Throwing.function((MessageResult messageResult) ->
+                messageFullViewFactory.fromMessageResults(ImmutableList.of(messageResult))).sneakyThrow())
+            .forEach(message -> storePreview(message.getId(), Preview.from(message.getPreview())));
+    }
+
+    private void storePreview(MessageId messageId, Preview preview) {
+        Mono.from(messagePreviewStore.store(messageId, preview)).block();
```

Comment:
Let's extends reactor usage to the preComputePreview methods IMO

Reply:
I agree the more we can defer the blocking, the better